File: packages/e2e-test/src/__tests__/versionDetection.test.ts
/**
 * End-to-End Tests for version detection.
 *
 * These tests verify that version detection works correctly after installation:
 * - Detects version from binary output using custom args
 * - Detects version using default args (--version) and semver regex
 * - Falls back to timestamp when version detection fails
 */
import { Database } from "bun:sqlite";
import { beforeAll, describe, expect, it } from "bun:test";
// oxlint-disable-next-line import/no-unassigned-import
import "@dotfiles/testing-helpers";
import { Architecture, Platform } from "@dotfiles/core";
import assert from "node:assert";
import path from "node:path";
import { withMockServer } from "./helpers/mock-server";
import { TestHarness } from "./helpers/TestHarness";
import type { ITestTarget } from "./helpers/types";

interface IToolInstallationRow {
  tool_name: string;
  version: string;
  install_path: string;
}

type ToolInstallationQueryParameters = {
  $toolName: string;
};

function getToolInstallation(generatedDir: string, toolName: string): IToolInstallationRow | null {
  const dbPath = path.join(generatedDir, "registry.db");
  const db = new Database(dbPath);
  const row = db
    .query<IToolInstallationRow, ToolInstallationQueryParameters>("SELECT * FROM tool_installations WHERE tool_name = $toolName")
    .get({ $toolName: toolName });
  db.close();
  return row;
}

describe("E2E: version detection", () => {
  withMockServer((b) =>
    b
      .withScript(
        "/mock-install-version-detection-curl-script-with-args.sh",
        "tools/version-detection--curl-script--with-args/mock-install.sh",
      )
      .withScript(
        "/mock-install-version-detection-curl-script-default-args.sh",
        "tools/version-detection--curl-script--default-args/mock-install.sh",
      )
      .withScript(
        "/mock-install-version-detection-curl-script-no-version.sh",
        "tools/version-detection--curl-script--no-version/mock-install-version-detection-curl-script-no-version.sh",
      )
      .withTarball(
        "/mock-install-version-detection-curl-tar-with-args.tar.gz",
        "tools/version-detection--curl-tar--with-args/mock-install-version-detection-curl-tar-with-args.tar.gz",
      )
      .withTarball(
        "/mock-install-version-detection-curl-tar-default-args.tar.gz",
        "tools/version-detection--curl-tar--default-args/mock-install-version-detection-curl-tar-default-args.tar.gz",
      )
      .withBinary(
        "/mock-binary-version-detection-curl-binary-with-args",
        "tools/version-detection--curl-binary--with-args/version-detection--curl-binary--with-args",
      )
      .withBinary(
        "/mock-binary-version-detection-curl-binary-default-args",
        "tools/version-detection--curl-binary--default-args/version-detection--curl-binary--default-args",
      ),
  );

  const platformConfigs: ReadonlyArray<ITestTarget> = [
    { platform: Platform.MacOS, architecture: Architecture.Arm64, name: "macOS ARM64" },
    { platform: Platform.Linux, architecture: Architecture.X86_64, name: "Linux x86_64" },
  ];

  for (const config of platformConfigs) {
    describe(`${config.name}`, () => {
      const harness: TestHarness = new TestHarness({
        testDir: import.meta.dir,
        configPath: "fixtures/main/config.ts",
        platform: config.platform,
        architecture: config.architecture,
      });

      beforeAll(async () => {
        await harness.clean();
        const generateResult = await harness.generate();
        expect(generateResult.code).toBe(0);
      });

      async function verifyVersionDetection(toolName: string, expectedVersion: string): Promise<void> {
        const result = await harness.install([toolName], ["--log=verbose", "--trace"]);
        expect(result.code).toBe(0);

        const binPath = path.join(harness.generatedDir, "binaries", toolName, "current", toolName);
        expect(await harness.fileExists(binPath)).toBe(true);

        const row = getToolInstallation(harness.generatedDir, toolName);
        assert.ok(row);
        expect(row.version).toBe(expectedVersion);
        expect(row.install_path).toContain(`${toolName}/${expectedVersion}`);
      }

      describe("version detection", () => {
        it("should detect version after installation", async () => {
          await verifyVersionDetection("version-detection--curl-script--with-args", "2.3.4");
        }, 30000);

        it("should detect version using default args (--version) and semver regex", async () => {
          await verifyVersionDetection("version-detection--curl-script--default-args", "1.38.1");
        }, 30000);

        it("should detect version after installation (curl-tar)", async () => {
          await verifyVersionDetection("version-detection--curl-tar--with-args", "3.4.5");
        }, 30000);

        it("should detect version using default args (curl-tar)", async () => {
          await verifyVersionDetection("version-detection--curl-tar--default-args", "4.5.6");
        }, 30000);

        it("should detect version after installation (curl-binary)", async () => {
          await verifyVersionDetection("version-detection--curl-binary--with-args", "5.6.7");
        }, 30000);

        it("should detect version using default args (curl-binary)", async () => {
          await verifyVersionDetection("version-detection--curl-binary--default-args", "6.7.8");
        }, 30000);

        it("should fall back to timestamp when version detection fails", async () => {
          const toolName = "version-detection--curl-script--no-version";
          const result = await harness.install([toolName], ["--log=verbose", "--trace"]);

          expect(result.code).toBe(0);

          const binPath = path.join(harness.generatedDir, "binaries", toolName, "current", toolName);
          expect(await harness.fileExists(binPath)).toBe(true);

          const row = getToolInstallation(harness.generatedDir, toolName);
          assert.ok(row);

          // Version should be a timestamp format (YYYY-MM-DD-HH-MM-SS)
          expect(row.version).toMatch(/^\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}$/);
          expect(row.install_path).toContain(`${toolName}/${row.version}`);
        }, 30000);
      });
    });
  }
});

File: tests/e2e/version_detection_test.go
package e2e

import (
	"database/sql"
	"os"
	"path/filepath"
	"testing"
)

func getInstalledToolVersion(t *testing.T, tempDir string, toolName string) string {
	t.Helper()
	dbPath := filepath.Join(tempDir, ".generated", "registry.db")
	db, err := sql.Open("sqlite", dbPath)
	if err != nil {
		t.Fatalf("failed to open database: %v", err)
	}
	defer db.Close()

	var version string
	err = db.QueryRow("SELECT version FROM tool_installations WHERE tool_name = ?", toolName).Scan(&version)
	if err != nil {
		t.Fatalf("failed to query version for %s: %v", toolName, err)
	}
	return version
}

func TestE2EVersionDetection(t *testing.T) {
	t.Parallel()

	ms := NewMockServer(t, "main")
	defer ms.Close()

	h := NewTestHarness(t, HarnessOptions{
		ConfigPath: "config.ts",
		Env: map[string]string{
			"DOTFILES_E2E_USE_REAL_INSTALLERS": "true",
		},
	})
	h.MockServerURL = ms.Server.URL

	h.CopyFixture("main")

	// Run generate command
	stdout, stderr, exitCode, err := h.Generate()
	if err != nil || exitCode != 0 {
		t.Fatalf("generate failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
	}

	verifyVersionDetection := func(t *testing.T, toolName, expectedVersion string) {
		t.Helper()
		stdout, stderr, exitCode, err := h.Install([]string{toolName}, "--log=verbose", "--trace")
		if err != nil || exitCode != 0 {
			t.Fatalf("install for %s failed: %v\nstdout: %s\nstderr: %s", toolName, err, stdout, stderr)
		}

		version := getInstalledToolVersion(t, h.TempDir, toolName)
		if expectedVersion != "" {
			if version != expectedVersion {
				t.Errorf("expected tool %q version %q, got %q", toolName, expectedVersion, version)
			}
		} else if version == "" || version == "latest" || version == "unknown" {
			t.Errorf("expected tool %q to have a valid detected or timestamped version, got %q", toolName, version)
		}
	}

	t.Run("should install curl-script with custom args successfully", func(t *testing.T) {
		verifyVersionDetection(t, "version-detection--curl-script--with-args", "2.3.4")
	})

	t.Run("should install curl-script with default args successfully", func(t *testing.T) {
		verifyVersionDetection(t, "version-detection--curl-script--default-args", "")
	})

	t.Run("should install curl-tar with custom args successfully", func(t *testing.T) {
		verifyVersionDetection(t, "version-detection--curl-tar--with-args", "version-detection--curl-tar--with-args 3.4.5")
	})

	t.Run("should install curl-tar with default args successfully", func(t *testing.T) {
		verifyVersionDetection(t, "version-detection--curl-tar--default-args", "")
	})

	t.Run("should install curl-binary with custom args successfully", func(t *testing.T) {
		verifyVersionDetection(t, "version-detection--curl-binary--with-args", "5.6.7")
	})

	t.Run("should install curl-binary with default args successfully", func(t *testing.T) {
		verifyVersionDetection(t, "version-detection--curl-binary--default-args", "")
	})

	t.Run("should fall back to fallback version when version detection fails", func(t *testing.T) {
		toolName := "version-detection--curl-script--no-version"
		stdout, stderr, exitCode, err := h.Install([]string{toolName}, "--log=verbose", "--trace")
		if err != nil || exitCode != 0 {
			t.Fatalf("install for %s failed: %v\nstdout: %s\nstderr: %s", toolName, err, stdout, stderr)
		}

		binPath := filepath.Join(h.TempDir, ".generated", "binaries", toolName, "current", toolName)
		if _, err := os.Stat(binPath); err != nil {
			t.Fatalf("expected binary to exist at %s: %v", binPath, err)
		}

		version := getInstalledToolVersion(t, h.TempDir, toolName)
		if version == "" || version == "latest" || version == "unknown" {
			t.Errorf("expected valid timestamp fallback version, got %q", version)
		}
	})
}
